plugins/artifacts/artifacts.go

```diff
@@ -0,0 +1,110 @@
+//go:generate go-composite-schema --unexported --required artifacts config-schema.yml generated_configschema.go
+
+// Package artifacts is responsible for uploading artifacts after builds
+package artifacts
+
+import (
+	"fmt"
+	"mime"
+	"path/filepath"
+
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+func (pluginProvider) NewPlugin(extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{}, nil
+}
+
+func (pluginProvider) ConfigSchema() runtime.CompositeSchema {
+	return runtime.NewEmptyCompositeSchema()
+}
+
+type plugin struct {
+	plugins.PluginBase
+}
+
+func (plugin) PayloadSchema() (runtime.CompositeSchema, error) {
+	return configSchema, nil
+}
+
+func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		payload:        *(options.Payload.(*config)),
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context *runtime.TaskContext
+	payload config
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
```

Comment:
I'm not sure if you want to do some validation here about the artifacts.  Like making sure artifact expires is <= task.expires otherwise when you go to create the artifact the queue will complain.